doc/manual/contract.md

```diff
@@ -0,0 +1,175 @@
+
+In Nit all the conditions of contract must be expressed using boolean expression in the same way as an assertion. Note that the potential routines that compose the condition of the contract should have no side effects.
+
+## Method/Attribute contract
+
+Methods and attributes (setter) can be associated with two types of contracts, entry contracts (expect) and exit contracts (ensure). Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfills the `ensure` condition on exit.
+
+Both are declared, between the definition of the signature followed by the keyword `is` and the block start keyword `do`.
+
+### Expect
+
+We define a precondition using the annotation `expect`. The following example shows how a precondition warns that a call to the `pop` method should be made on a non-empty stack.
+
+~~~
+	fun pop
+	is
+		expect(not_empty)
+	do
+		# Remove the last item on the stack
+	end
+~~~
+
+### Ensure
+
+We define a postcondition using the annotation `ensure`. Here is an example with a `push` method which adds the element received as an argument (`object`) at the top of the stack. It garantees to the caller that the object will be added at the top of the stack `self [size] == object`.
+
+~~~
+	fun push(object: T)
+	is
+		ensure(self[size] == object)
+	do
+		# Add the given argument `object` on top of the stack
+	end
+~~~
+
+The `ensure` contract offers the possibility of referring to the return value of the method, using the keyword `result`. Example the `get_at` method which returns the object which is stored at the index received as an argument (`index`). It guarantees the caller that the returned element represented by `result` corresponds to the element stored in the stack with the given index `result == self [index]`.
+
+~~~
+	fun get_at(index: Int): T
+	is
+		ensure(result == self[index])
+	do
+		# Return the element at the given index.
+	end
+~~~
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, using the pop example, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
+
+~~~
+	fun pop(object: T)
+	is
+		ensure(old(size) - 1 == size)
+	do
+		# Remove the last item on the stack
+	end
+~~~
+
+### Additional informations
+
+* When defining a `ensure` or `expect` contract on an attribute, only the automatically generated write property will use the contract.
+
+~~~
+	var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
+~~~
+
+* The `ensure` or `expect` contracts are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
```

Comment:
Always present them in the same order as to not confuse people of which is the keyword for before and which is the one for after.